janus_core/helpers/utils.py

```diff
@@ -229,3 +231,89 @@ def dict_remove_hyphens(dictionary: dict) -> dict:
         if isinstance(value, dict):
             dictionary[key] = dict_remove_hyphens(value)
     return {k.replace("-", "_"): v for k, v in dictionary.items()}
+
+
+def results_to_info(
+    struct: Atoms,
+    *,
+    properties: Collection[Properties] = (),
+    invalidate_calc: bool = False,
+) -> None:
+    """
+    Copy or move MLIP calculated results to Atoms.info dict.
+
+    Parameters
+    ----------
+    struct : Atoms
+        Atoms object to copy or move calculated results to info dict.
+    properties : Collection[Properties]
+        Properties to copy from results to info dict. Default is ().
+    invalidate_calc : bool
+        Whether to remove all calculator results after copying properties to info dict.
+        Default is False.
+    """
+    if not properties:
+        properties = get_args(Properties)
+
+    if struct.calc:
+        # Set default architecture from calculator name
+        arch = struct.calc.parameters["arch"]
+        struct.info["arch"] = arch
+
+        for key in properties & struct.calc.results.keys():
+            tag = f"{arch}_{key}"
+            value = struct.calc.results[key]
+            if key == "forces":
+                struct.arrays[tag] = value
+            else:
+                struct.info[tag] = value
+
+        # Remove all calculator results
+        if invalidate_calc:
+            struct.calc.results = {}
+
+
+def output_structs(
+    images: MaybeSequence[Atoms],
+    *,
+    set_info: bool = True,
+    write_results: bool = False,
+    properties: Collection[Properties] = (),
+    invalidate_calc: bool = False,
+    **kwargs,
+) -> None:
+    """
+    Copy or move calculated results to Atoms.info dict and/or write structures to file.
+
+    Parameters
+    ----------
+    images : MaybeSequence[Atoms]
+        Atoms object or a list of Atoms objects to interact with.
+    set_info : bool
+        True to set info dict from calculated results. Default is True.
+    write_results : bool
+        True to write out structure with results of calculations. Default is False.
+    properties : Collection[Properties]
+        Properties to copy from calculated results to info dict. Default is ().
+    invalidate_calc : bool
+        Whether to remove all calculator results after copying properties to info dict.
+        Default is False.
+    **kwargs
+        Keyword arguments passed to ase.io.write.
+    """
+    if isinstance(images, Atoms):
+        images = (images,)
+
+    if set_info:
+        for image in images:
+            results_to_info(
+                image, properties=properties, invalidate_calc=invalidate_calc
+            )
+    else:
+        # Label architecture even if not copying results to info
+        for image in images:
+            if image.calc:
```

Comment:
Is this likely to be called without an `image.calc`? I assume the answer is yes.

Reply:
Not particularly likely, actually.

None of the cases where it's called at the moment should be False here, since attaching a calculator is generally the first thing we do before anything else, and there's no reason not to use `ase.io.write` (or write at all) if we haven't calculated anything yet.

Arguably this could be a value error `if not image.calc`, but I think I wrote it like this to make it generally a safe swap for `ase.io.write`, even if practically it shouldn't be needed usually.